### ISSUE TITLE
Face recognitions is not working properly in my project

### ISSUE THREAD

**Issue by Rohit-B-Kadam** [justadudewhohacks/face-api.js#264]
I am doing a project in angular7 and electronjs. In my project I want to detect and recognise the face from given image. But I am facing a problem in face recognition it's not recognizing the face properly 
![Screenshot from 2019-04-05 16-27-44](https://user-images.githubusercontent.com/34840376/55624393-505ca980-57c3-11e9-9c56-88534dfad676.png)

But went I try with your example project face was recognizing properly.
![Screenshot from 2019-04-05 16-29-46](https://user-images.githubusercontent.com/34840376/55624531-df69c180-57c3-11e9-8061-f6c99f90c572.png)
![Screenshot from 2019-04-05 16-29-50](https://user-images.githubusercontent.com/34840376/55624532-e0025800-57c3-11e9-9f93-b2eef6c8b450.png)

**Comment by erenaud3** [comment]
Hi !

It is kind of hard to tell what is wrong in your project if we don't see your code. Could you add a code snippet with the involved code ?

What do you want to achieve exaclty in first screenshot ? If I understand correctly, the first image is used to get face descriptor in order to populate a Face Matcher and you want to recognize this person in second image, am I right ?

According to the few information we have, you are maybe not adding correctly face descriptor from first image to your face matcher.

```
_currentLabeledDescriptor = new faceapi.LabeledFaceDescriptors("person1", [_descriptorsFromFirstPicture])
_localFaceMatcher  = new faceapi.FaceMatcher([_currentLabeledDescriptor])
```

Double check that _descriptorsFromFirstPicture is a Float32Array and that you have "logical" values (i,e, floats with a lot of digits and values in range [-1;1]).

**Comment by Rohit-B-Kadam** [comment]
yes, I want to recognise person from first image in Second image. Your App recognise it properly but my code is not recognising it.

First Image Facedetection code( .ts ):

    const firstFaceDescription = await faceapi
      .detectSingleFace(img)
      .withFaceLandmarks().withFaceDescriptor()

    if (!isNullOrUndefined(firstFaceDescription)) 
    {
      const detectionsForSize = faceapi.resizeResults(firstFaceDescription,
        { width: profileImgRef.width, height: profileImgRef.height })
      
      const canvas = this.canvas1.nativeElement;
      canvas.width = profileImgRef.width;
      canvas.height = profileImgRef.height;

      const detectionsArray = detectionsForSize.detection

      const singleLabelDiscription = new faceapi.LabeledFaceDescriptors("person1", [firstFaceDescription.descriptor])
      this.faceMatcher = new faceapi.FaceMatcher(singleLabelDiscription)

      faceapi.drawDetection(canvas, detectionsArray, { withScore: true, withClassName: true })
    }


Second image detection code
     
     const fullFaceDescriptions = await faceapi
                      .detectAllFaces(img)
                      .withFaceLandmarks().withFaceDescriptors()

    if (!isNullOrUndefined(fullFaceDescriptions)) 
    {
      console.log(fullFaceDescriptions);
      
      const detectionsForSize = faceapi.resizeResults(fullFaceDescriptions, 
        { width: groupImgRef.width, height: groupImgRef.height })
      
      const canvas = this.canvas2.nativeElement;
      canvas.width = groupImgRef.width;
      canvas.height = groupImgRef.height;
      
      const detectionsArray = detectionsForSize.map(fd => fd.detection)

      // Face recognition
      const results = fullFaceDescriptions.map(fd => this.faceMatcher.findBestMatch(fd.descriptor))

      console.log(results)

      const boxesWithText = results.map((bestMatch, i) => {
        const box = detectionsArray[i].box
        const text = bestMatch.toString()
        const boxWithText = new faceapi.BoxWithText(box, text)
        return boxWithText
      })

      faceapi.drawDetection(canvas, boxesWithText)
    }

![Screenshot from 2019-04-08 23-52-28](https://user-images.githubusercontent.com/34840376/55747322-9cc11700-5a59-11e9-8573-0ea1bfae61dd.png)

One more question: How many face descriptor of each person will required for recognition that person

**Comment by justadudewhohacks** [comment]
Since both distances are 0, odds are, that you are comparing the exact same image against eachother.

> One more question: How many face descriptor of each person will required for recognition that person
One descriptor is sufficient.

**Comment by phoenixsue** [comment]
justadudewhohacks He is right. There's definitely a big problem with your library. I computed the face descriptors(as in your example) for DIFFERENT images of DIFFERENT( and i really mean different) people and guess what, in some of them i got the exact same face descriptor. But the strange thing is that the descriptor is THE EXACT ONE showed by the indian man in the picture above, which is
-0.036157943308353424,0.11451219767332077,0.08395102620124817,-0.015166843309998512,-0.06600701808929443,-0.042419757694005966,-0.0490785650908947,0.008687270805239677,0.14321738481521606,-0.025509962812066078,0.2366054803133011,0.001498898956924677,-0.23820878565311432,-0.046752505004405975,0.0028616711497306824,0.08502069860696793,-0.14962242543697357,-0.06591969728469849,-0.10743647068738937,-0.12046073377132416,0.033088553696870804,0.00343224941752851,0.018290642648935318,0.009096329100430012,-0.09878937155008316,-0.2763112187385559,-0.07352373749017715,-0.1335674375295639,0.08969920873641968,-0.19571422040462494,0.012862971052527428,0.021690549328923225,-0.13443519175052643,-0.05735789239406586,-0.02766452543437481,0.02025718241930008,0.0033780643716454506,-0.08854390680789948,0.15200215578079224,-0.03676857054233551,-0.14023277163505554,-0.053991273045539856,-0.010764091275632381,0.2646256685256958,0.18705308437347412,0.044662196189165115,0.00966082513332367,-0.043214693665504456,0.06518538296222687,-0.2766679525375366,0.02771242894232273,0.1624835729598999,0.02950565330684185,0.12462048977613449,0.06644287705421448,-0.1700247824192047,0.0408409908413887,0.08843832463026047,-0.1381165087223053,0.03919530287384987,0.0404399111866951,-0.09420391917228699,-0.036383725702762604,-0.11028037220239639,0.21431322395801544,0.06784548610448837,-0.08882343769073486,-0.12961089611053467,0.10797084122896194,-0.14936871826648712,-0.033338963985443115,0.11759535223245621,-0.11118640750646591,-0.16549569368362427,-0.22307536005973816,0.10256917774677277,0.3594265282154083,0.19990438222885132,-0.1451532542705536,0.036446064710617065,-0.1241573691368103,-0.01249675266444683,0.025015369057655334,0.023310799151659012,-0.06697430461645126,0.016636021435260773,-0.09167926013469696,0.05546090751886368,0.1520289182662964,-0.07664483040571213,0.02443789131939411,0.18324145674705505,-0.05131388455629349,0.03371473029255867,-0.0032856762409210205,0.01570814847946167,-0.08626414090394974,0.053548894822597504,-0.06702996790409088,0.01747414842247963,0.12703940272331238,-0.1353486031293869,0.027928803116083145,0.06764037907123566,-0.17105481028556824,0.10737797617912292,-0.004590324591845274,-0.020738571882247925,0.03764114901423454,0.005512760952115059,-0.06117083877325058,-0.036734696477651596,0.23441985249519348,-0.26262736320495605,0.24768288433551788,0.23473739624023438,0.03640888258814812,0.14327383041381836,0.039577461779117584,0.1486268937587738,-0.04878677800297737,-0.06605279445648193,-0.1222165897488594,-0.028063280507922173,0.012238509021699429,-0.014763789251446724,-0.04189039394259453,-0.003825186751782894
How can 2 different images of two different persons have the EXACT same descriptor ?!?

**Comment by justadudewhohacks** [comment]
It is hard to say what's going wrong without knowing your inputs. Can you share the images you are facing the issue with (reference and query image).

**Comment by Rohit-B-Kadam** [comment]
> Since both distances are 0, odds are, that you are comparing the exact same image against eachother.
>

Both image are not same. 

And I have notice that this three face description has value(Float32Array) exactly same. How? 

my code
https://github.com/Rohit-B-Kadam/Memento/tree/master/src/app/face-detect

Image
https://github.com/Rohit-B-Kadam/Memento/tree/master/Data

one warning i have seen
  tf-core.esm.js:17 performance warning: READ-usage buffer was read back without waiting on a fence. This caused a graphics pipeline stall.

**Comment by phoenixsue** [comment]
@justadudewhohacks i created a repository here: https://github.com/phoenixsue/face-api-issues
I included 4 images, all of which produce the infamous descriptor i showed above. 
Maybe that descriptor is used for initializing the array but then something happens and  instead of returning an error that's what we get. As for the code, i' m using electron 4.1.0. I downloaded the zip file of your library from this repository and included in the index.html file like this: 
<script type="text/javascript" src="js/face-api.js/dist/face-api.js"></script>
The rest of the code is shown in my previous comment.
Please try to solve this problem otherwise your library is unusable

**Comment by phoenixsue** [comment]
Hmmm, i tried the same images in a simple web page with only your library loaded,  no server, no electron, no nothing and i got the right descriptors.  So i think that the problem is with electron. I' m confused...

**Comment by justadudewhohacks** [comment]
@Rohit-B-Kadam @phoenixsue, the images you have shared work fine on a web application. Seems to be related to electron then. A similar issue with electron has been reported in #262. Which backend are you guys using on the web (webgl?, cpu?) and what backend in the electron app?

@phoenixsue thanks for setting up an example. I will check it out on the weekend. If we figure out the ops, that are causing different results between web and electron we can certainly file this issue at tfjs.

**Comment by phoenixsue** [comment]
"Which backend are you guys using on the web (webgl?, cpu?) and what backend in the electron app?"
I don' t understand the question(i'm an amateur programmer, you know). What do you mean by backend?
Can you explain please?

**Comment by justadudewhohacks** [comment]
You can check `faceapi.tf.getBackend()`, which will tell you whether the cpu or webgl backend is used. The webgl backend runs the ops on your gpu.

**Comment by phoenixsue** [comment]
it says webgl. What does it mean?

**Comment by justadudewhohacks** [comment]
@phoenixsue I tried out your example repo and indeed there is an issue with utilizing the webgl backend in electrons renderer thread. If I run the example on the CPU, e.g. by calling `faceapi.tf.setBackend('cpu')` the descriptors are calculated as expected.

Now this is kind of a hairy situation. If we are going to file an issue at tfjs for that, we might have to figure out the exact cause of the issue, e.g. which ops are causing the divergence between the browser webgl backend and the webgl backend of the electron renderer thread. 

One approach would be to compare the outputs of each layer of the face recognition net. Since the face-recognition net is the only net utilizing regular convolutions, the issue might maybe reside in tf.conv2d.

**Comment by phoenixsue** [comment]
@Vincent Mühler Thank you for your effort. I checked too and the descriptors seem to be the right ones BUT it' s painfully SLOW on the CPU :((. I don' t know how to help you. I understand nothing about the details and machine learning in general. Why don' t you just show them the example you have worked with and tell them what the problem is? They could fix it in no time. No?

**Comment by thirukumars** [comment]
i am using face-api for live browser recognisation using mtcnn,
i follow the instruction in mtcnn documentation but when i try to run i got this error
  

Uncaught (in promise) TypeError: faceapi.drawDetection is not a function
    at onPlay (script2.js:37)

**Comment by thirukumars** [comment]
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>FaceDetect</title>
  
  <script defer src="face-api.min.js"></script>
  <script defer src="script2.js"></script>
 
  
  
  <style>
    body {
      margin: 0;
      padding: 0;
      width: 100vw;
      height: 100vh;
      display: flex;
      justify-content: center;
      align-items: center;
      background-color: azure;
    }
    
  </style>
</head>
<body>
  <div style="position: relative" class="margin">
    <video onplay="onPlay(this)" id="inputVideo" autoplay muted></video>
    <canvas id="overlay" />
  </div>
</body>
</html>


this my HTML

**Comment by thirukumars** [comment]
[index.txt](https://github.com/justadudewhohacks/face-api.js/files/3455241/index.txt)
 this is html

**Comment by thirukumars** [comment]
[script.txt](https://github.com/justadudewhohacks/face-api.js/files/3455247/script.txt)

this is the script file

**Comment by thirukumars** [comment]
any one please help with the solution

**Comment by neumartin** [comment]
Maybe in electron is best to use the main thread to use TF and not the render thread.
In main thread I think we can use GPU backend with an Nvidia GPU.
What do you think @justadudewhohacks ?

**Comment by ImposibleScience** [comment]
Hi, I am experiencing an issue. When I Downloaded the files and ran it I got a blank white screen pls help

**Comment by erenaud3** [comment]
@ImposibleScience  "Hi ! Can you help me about a problem I give no information about ?"

If you want help, open a new issue first. How is your problem related to current issue in any way ??

Then, consider making a REAL description of your problem.
